[5.2] mod_redirect - Admin Module to display Redirects: Links

pe7er opened the pull request:
### Summary of Changes
This is PR creates **mod_redirect**, a new **Admin Module** that displays "**Redirects: Links**", the links that are displayed in the "Redirects" component. 

**Example output**:
![dashboard](https://github.com/user-attachments/assets/cb42f8a2-05a3-46c2-9449-95072b3129dc)


### Background
Broken links have a negative impact on the user experience and also harm SEO. 
Joomla has in the core an excellent Redirects component that administrators can use to manually create "301 redirects" for faulty links. 

Administrators can manage faulty links via: System > (under Manage) Redirects. 
It makes sense to redirect the most visited unredirected faulty links to working pages.
To display these links in the Redirects component:
- Go to [Filter Options]
- Set State to "Disabled"
- Sort the table by "404 Hits descending"

![redirects-links-component](https://github.com/user-attachments/assets/9653068c-37a2-4f45-a5db-c974f1bff81e)

### Admin Module: mod_redirect
To simplify the process for admins, this new Redirect Admin Module displays a quick overview of faulty links that can be added to a Dashboard. 

By default, the module displays:
- The top 5 redirects links
- Links with a "disabled" state
- Links with an HTTP status code of "301"
- The list is sorted by the number of hits in descending order, with the most frequently hit links at the top.

This way it's easier to see which faulty links in the website need to be redirected.  

### Installation
Add the PR to your local environment (e.g. using the Joomla Patchtester Component)

**Install the module** via: System > **Discover** > install "Redirects: Links".
- When this module is added to the Joomla core, we need to update the #__extensions table via /installation/sql/mysql/base.sql + /installation/sql/postgresql/base.sql.
- This module should NOT be added to the dashboard by default because the Redirect Plugin is not enabled by default.

![extension-discover](https://github.com/user-attachments/assets/5b01551d-51d5-4ae6-841f-8b82e406ebf7)

**Create a New module** via: Content > Administrator Modules, New 

![module-new](https://github.com/user-attachments/assets/e7b79935-ce5d-49ea-b06a-9e3853bc3691)

**Configure the Module**, add a Title, set the Position to "cpanel" and status to "Published,"
 
![module-config](https://github.com/user-attachments/assets/b7aa5148-f6b5-44e7-a865-f470f5af6bc8)

### Sample data
**To test** the configuration and output of **the module**, the **Redirects Component needs to have some links**. 

You can create your own faulty links within Joomla:
- The Redirect Plugin must be enabled
- In the Redirect component, the 'Collect URLs' option should be enabled
- Visit faulty links on the front-end to generate entries in the Redirects component

Or you can use the following **SQL sample data** as "Redirects: Links".
Run the following SQL in your database (e.g. using phpMyAdmin) to create sample records.
You'll need to change the #__ prefix with your Joomla's database prefix.

```sql
INSERT INTO `#__redirect_links` (`old_url`, `new_url`, `referer`, `comment`, `hits`, `published`, `created_date`, `modified_date`, `header`) VALUES
('https://local.test.nl/faulty-link', NULL, '', '', 101, 0, '2024-08-08 08:13:29', '2024-08-11 08:10:02', 301),
('https://local.test.nl/faulty-link-archived', NULL, '', '', 1, 2, '2024-08-05 08:10:08', '2024-08-11 08:10:08', 301),
('https://local.test.nl/faulty-link-trashed', NULL, '', '', 1, -2, '2023-08-06 08:10:15', '2024-08-11 08:10:15', 301),
('https://local.test.nl/faulty-link-with-xss-attempt&param=<script>alert(\'xss\')</script>', NULL, '', '', 1001, 0, '2023-07-23 12:11:26', '2024-08-11 08:11:26', 301),
('https://local.test.nl/faulty-link-disabled', NULL, '', '', 1, 0, '2021-01-21 18:12:45', '2024-08-11 08:12:45', 301),
('https://local.test.nl/faulty-link-redirected', 'https://local.test.nl/', '', 'redirect to home', 13, 1, '2024-08-08 08:13:29', '2024-08-11 08:13:29', 301),
('https://local.test.nl/another-faulty-link', NULL, '', '', 1, 0, '2024-08-11 08:13:47', '2024-08-11 08:13:47', 301),
('https://local.test.nl/yet-another-faulty-link', NULL, '', '', 1, 0, '2023-01-01 01:02:03', '2024-08-11 08:13:52', 301),
('https://local.test.nl/what-is-wrong-at-this-site', NULL, '', '', 11, 0, '2024-08-11 08:14:07', '2024-08-11 08:14:07', 301),
('https://local.test.nl/pe7er-was-here', NULL, '', '', 93258, 0, '2024-08-11 10:10:10', '2024-08-11 08:14:15', 301);
```

### Testing Instructions
- Test changes in the Redirect component. Modify the state of some links. Check how these changes reflect in the "Redirects: Links" Module.
- Test changes in the module configuration. Check if those are displayed correctly on the Dashboard
- Test that the displayed links are sanitized against vulnerabilities like JavaScript URL injection.

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: **Documentation needs to be created. I'll create documentation after this feature has been merged**
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

pe7er committed:
[5.2] mod_redirect - Admin Module to display Redirects: Links

richard67 commented:
> When this module is added to the Joomla core, we need to update the #__extensions table via /installation/sql/mysql/base.sql + /installation/sql/postgresql/base.sql.

@pe7er This will not be enough. It will also need an update SQL script for each database type for adding the module to the extensions table, and as long both (update SQL and base.sql) is not done, this PR is not ready. We cannot merge it when it needs a discovery installation to install the extension.

brianteeman commented:
It also needs to be added to the list of core extensions
https://github.com/joomla/joomla-cms/blob/7aaa9d3f99f40dbfb535edae94b9fad6f77cbb1f/libraries/src/Extension/ExtensionHelper.php#L107

richard67 commented:
If necessary I can help with the missing stuff, but during the week only in the evenings after work, if I am not too tired. Or next weekend.

brianteeman commented:
This needs a LOT of further work before it is useful (personally I dont see the point as I dont see what benefit it offers and it could easily be something on the JED )

For example
1. If the redirect system plugin is not enabled there is no notification of this in the module and it will always be empty
2. If the redirect system plugin is enabled but the collect urls option is not enabled  there is no notification of this in the module and it will always be empty
3. The module has an option to _filter_ on the http status but this option is only applicable if the component has advanced mode enabled
4. why would you ever want to display a redirect link that is trashed

pe7er committed:
Updated SQL script + removed ordering options #43908

brianteeman commented on installation/sql/postgresql/base.sql:
```diff
@@ -673,7 +674,8 @@ INSERT INTO "#__modules" ("id", "asset_id", "title", "note", "content", "orderin
 (108, 85, 'Privacy Status', '', '', 1, 'cpanel-privacy', NULL, NULL, 1, 'mod_privacy_status', 1, 1, '{"layout":"_:default","moduleclass_sfx":"","cache":1,"cache_time":900,"cachemode":"static","style":"0","module_tag":"div","bootstrap_size":"12","header_tag":"h2","header_class":""}', 1, '*'),
-(109, 96, 'Guided Tours', '', '', 1, 'status', NULL, NULL, 1, 'mod_guidedtours', 1, 1, '', 1, '*');
+(109, 96, 'Guided Tours', '', '', 1, 'status', NULL, NULL, 1, 'mod_guidedtours', 1, 1, '', 1, '*'),
+(110, 100, 'Redirects: Links', '', '', 1, 'cpanel', NULL, NULL, 0, 'mod_redirect', 1, 1, '{\"state\":\"0\",\"http_status\":\"301\",\"ordering\":\"a.hits DESC\",\"count\":5,\"layout\":\"_:default\",\"moduleclass_sfx\":\"\",\"automatic_title\":0,\"cache\":1,\"cache_time\":900,\"cachemode\":\"static\",\"module_tag\":\"div\",\"bootstrap_size\":\"0\",\"header_tag\":\"h3\",\"header_class\":\"\",\"style ...
```
i dont think you need to escape the " as it isnt done in the rest of the sql

pe7er replied:
thanks

brianteeman commented on installation/sql/mysql/base.sql:
```diff
@@ -649,8 +650,8 @@ INSERT INTO `#__modules` (`id`, `asset_id`, `title`, `note`, `content`, `orderin
-(109, 96, 'Guided Tours', '', '', 1, 'status', NULL, NULL, 1, 'mod_guidedtours', 1, 1, '', 1, '*');
-
+(109, 96, 'Guided Tours', '', '', 1, 'status', NULL, NULL, 1, 'mod_guidedtours', 1, 1, '', 1, '*'),
+(110, 100, 'Redirects: Links', '', '', 1, 'cpanel', NULL, NULL, 0, 'mod_redirect', 3, 1, '{\"state\":\"0\",\"http_status\":\"301\",\"ordering\":\"a.hits DESC\",\"count\":5,\"layout\":\"_:default\",\"moduleclass_sfx\":\"\",\"automatic_title\":0,\"cache\":1,\"cache_time\":900,\"cachemode\":\"static\",\"module_tag\":\"div\",\"bootstrap_size\":\"0\",\"header_tag\":\"h3\",\"header_class\":\"\",\"style ...
```
i dont think you need to escape the " as it isnt done in the rest of the sql

sandewt commented on administrator/modules/mod_redirect/src/Helper/RedirectHelper.php:
```diff
@@ -0,0 +1,58 @@
+     * @throws  \Exception
+     * @since   __DEPLOY_VERSION__
+     */
+    public function getRedirectLinks(&$params)
```
```suggestion
    public function getRedirectLinks(Registry $params, AdministratorApplication $app)
```
and so on

pe7er replied:
thanks

sandewt commented on administrator/modules/mod_redirect/mod_redirect.xml:
```diff
@@ -0,0 +1,132 @@
+	<config>
+		<fields name="params">
+			<fieldset name="basic" addfieldprefix="Joomla\Component\Redirect\Administrator\Field">
+
```
```suggestion
```

sandewt commented on administrator/modules/mod_redirect/mod_redirect.xml:
```diff
@@ -0,0 +1,132 @@
+					type="RedirectStatus"
+					label="JSTATUS"
+					default="0"
+					class="js-select-submit-on-change"
+				>
+					<option value="">JOPTION_SELECT_PUBLISHED</option>
```
```suggestion
					class="js-select-submit-on-change"
				        >
					<option value="">JOPTION_SELECT_PUBLISHED</option>
```
And so on

pe7er replied:
thanks

sandewt commented on administrator/modules/mod_redirect/mod_redirect.xml:
```diff
@@ -0,0 +1,132 @@
+	<license>GNU General Public License version 2 or later; see LICENSE.txt</license>
+	<authorEmail>admin@joomla.org</authorEmail>
+	<authorUrl>www.joomla.org</authorUrl>
+	<version>1.0.0</version>
```
```suggestion
	<version>5.2.0</version>
```

pe7er committed:
removed escapes, added typehint, reformatted xml

pe7er committed:
Merge branch '5.2-dev' into features/mod_redirect

pe7er committed:
fixed issue with Postgresql autoincrement

pe7er committed:
Merge branch '5.2-dev' of github.com:pe7er/joomla-cms into features/mod_redirect

pe7er committed:
Merge branch 'features/mod_redirect' of github.com:pe7er/joomla-cms into features/mod_redirect

pe7er committed:
reverted back to original postgresql/base.sql

pe7er committed:
added mod_redirect to postgresql/base.sql

brianteeman commented:
Will you be addressing the oustanding issues?

pe7er committed:
changed authorization, state filter, th scope

pe7er commented:
Yes

> If the redirect system plugin is not enabled there is no notification of this in the module and it will always be empty
> If the redirect system plugin is enabled but the collect urls option is not enabled there is no notification of this in the module and it will always be empty

IMHO not necessary. Check the out put of other core modules when they don't contain records:
Recently Added Articles, Privacy Dashboard and Popular Articles all display: "No Articles have been created yet."

If you are not using the Articles (because of using only 3rd party webshop or page builder extension),
or no-one every used the Privacy function, those won't change either.

> The module has an option to filter on the http status but this option is only applicable if the component has advanced mode enabled

I don't think that there's an option to filter the "Redirect" form type options on the mode of the Redirects component.  

> why would you ever want to display a redirect link that is trashed

true. fixed

> it doesnt take any acl into account when providing the links to edit the url in the redirect component

Actually, it did. The whole module was only visible by users with core.admin rights:
```php
public function dispatch()
    {
        if (!$this->getApplication()->getIdentity()->authorise('core.admin')) {
            return;
        }
```

I've change that so that the whole module is only visible for users that have edit rights in the redirect component:
```php
        if (!Factory::getApplication()->getIdentity()->authorise('core.edit', 'com_redirect')) {
```


> it is very confusing to have sort options for the module for data that is not presented in the table (for example what use is it to sort by referring page if the module never displays that info)

true. fixed

> Accessibility - the main column in a row (in this case the old_url) should be a th with a scope of row

thanks. fixed

> Update sql missing

Was already fixed.

brianteeman commented:
> > Update sql missing

> Was already fixed.

No it is not

brianteeman commented:
> > If the redirect system plugin is not enabled there is no notification of this in the module and it will always be empty
> > If the redirect system plugin is enabled but the collect urls option is not enabled there is no notification of this in the module and it will always be empty

> IMHO not necessary. Check the out put of other core modules when they don't contain records:
> Recently Added Articles, Privacy Dashboard and Popular Articles all display: "No Articles have been created yet."

> If you are not using the Articles (because of using only 3rd party webshop or page builder extension),
or no-one every used the Privacy function, those won't change either.

Not true - just because the plugin is not enabled and/or the collect urls is not enabled does not mean that there are no redirects 


-- still not seen anyone say that they wanted this module

pe7er commented:
> -- still not seen anyone say that they wanted this module

Not every Joomla user has a github account or dares to participate in discussions there.

I have spoken to quite a few Joomla users who think this is a welcome feature for the Joomla core.

pe7er commented:
> > > Update sql missing
> 
> > Was already fixed.
> 
> No it is not

Yes it was. Please check the commit before your "Will you be addressing the oustanding issues?" question:
https://github.com/joomla/joomla-cms/pull/43908/commits/1f3c7331992dce6ec1c39943d11760dedb5030fb

brianteeman commented:
@pe7er that is not for updates!!! that is for new installs only. https://github.com/joomla/joomla-cms/tree/5.2-dev/administrator/components/com_admin/sql/updates

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on 0dee7fa5f72bab768ab78f3f260ef8472370db96<br /><br />Does not work on any updated site<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43908">issues.joomla.org/tracker/joomla-cms/43908</a>.</sub>

pe7er committed:
added update sql for existing sites

brianteeman commented on administrator/components/com_admin/sql/updates/mysql/5.2.0-2024-08-15.sql:
```diff
@@ -0,0 +1,12 @@
+(0, 'mod_redirect', 'module', 'mod_redirect', '', 1, 1, 0, 0, 1, '{}', '{}', '', 0, 0);
+
+INSERT INTO `#__modules` (`title`, `note`, `content`, `ordering`, `position`, `checked_out`, `checked_out_time`, `publish_up`, `publish_down`, `published`, `module`, `access`, `showtitle`, `params`, `client_id`, `language`) VALUES
+('Redirects: Links', '', '', 1, 'cpanel', NULL, NULL, '2024-08-15 14:46:26', NULL, 0, 'mod_redirect', 3, 1, '{\"state\":\"0\",\"http_status\":\"301\",\"ordering\":\"a.hits DESC\",\"count\":5,\"layout\":\"_:default\",\"moduleclass_sfx\":\"\",\"automatic_title\":0,\"cache\":1,\"cache_time\":900,\"cachemode\":\"static\",\"module_tag\":\"div\",\"bootstrap_size\":\"0\",\"header_tag\":\"h3\",\"header_c ...
```
You dont need to set a value for publish_up - it should be NULL as you have done on the install sql

brianteeman replied:
As before you dont need to escape the " in the sql

brianteeman replied:
Is it intentional that the module is unpublished?

brianteeman replied:
Is it a problem that you dont set an asset_id ?

brianteeman commented:
I have tested this item :red_circle: unsuccessfully on 503b7d5916a62b0b1c6549b4a421232475a0c9cd<br /><br />Tested again with both an update and a clean install of this branch

In both cases the admin module is created and is listed in the module manager for admin modules and is unpublished. (not sure if that is intentional)

After publishing the module it is still not displayed at all on the admin control panel<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43908">issues.joomla.org/tracker/joomla-cms/43908</a>.</sub>

brianteeman commented:
https://github.com/user-attachments/assets/97a14b52-b34d-448b-ad06-c3667271b69d

brianteeman commented:
Might be logical but its not consistent with all other admin modules. Anyway it doesnt appear when published as shown in the video

sandewt commented:
>Install the module via: System > Discover > install "Redirects: Links".

![Schermafbeelding 2024-08-17 124831](https://github.com/user-attachments/assets/d97eb776-5478-4776-9c6d-5308cf97848d)

sandewt commented on administrator/modules/mod_redirect/src/Helper/RedirectHelper.php:
```diff
@@ -0,0 +1,58 @@
+     * @throws  \Exception
+     * @since   __DEPLOY_VERSION__
+     */
+    public function getRedirectLinks(Registry &$params)
```
```suggestion
    public function getRedirectLinks(Registry $params)
```

richard67 commented on installation/sql/postgresql/base.sql:
```diff
@@ -122,9 +122,10 @@ INSERT INTO "#__assets" ("id", "parent_id", "lft", "rgt", "level", "name", "titl
 (98, 1, 179, 180, 1, 'com_scheduler.task.2', 'com_scheduler.task.2', '{}'),
-(99, 1, 181, 182, 1, 'com_scheduler.task.3', 'com_scheduler.task.3', '{}');
+(99, 1, 181, 182, 1, 'com_scheduler.task.3', 'com_scheduler.task.3', '{}'),
+(100, 18, 183, 184, 2, 'com_modules.module.110', 'Redirects: Links', '{}');
```
In the base.sql for MySQL you don't insert an asset, but here you do. Why this inconsistency?
And if you insert an asset you also have to increase the `rgt` value of the root item from 183 to 185 here: https://github.com/joomla/joomla-cms/blob/503b7d5916a62b0b1c6549b4a421232475a0c9cd/installation/sql/postgresql/base.sql#L34

pe7er commented:
Thanks @richard67 @brianteeman @sandewt for testing and improvements.

As I wasn't able to convince all the other maintainers to include it in the core at this time, I decided to make this Joomla 5 version available as installable module: https://github.com/pe7er/mod_redirect

sandewt commented:
>As I wasn't able to convince all the other maintainers to include it in the core at this time

To avoid disappointments, there should be some kind of checklist with minimum requirements that an extension must meet to be included in the Joomla core. At the moment, it seems to depend too much on the maintainers. Don't get me wrong, I'm not saying anything here to the detriment of the maintainers. I know of a case where the designer turned away from Joomla in frustration. The more objective the criteria for admission to the core, the better.

Is there such a checklist available now? If not, is it possible that this will come?
